addons/stock/models/stock_quant.py

```diff
@@ -777,7 +777,7 @@ def _gather(self, product_id, location_id, lot_id=None, package_id=None, owner_i
         domain, order = self._get_removal_strategy_domain_order(domain, removal_strategy, qty)
         if self.ids:
             sort_key = self._get_removal_strategy_sort_key(removal_strategy)
-            res = self.filtered_domain(domain).sorted(key=sort_key[0], reverse=sort_key[1])
+            res = (self & self.with_context(prefetch_fields=False).search(domain)).sorted(key=sort_key[0], reverse=sort_key[1])
```

Comment:
- At this point in time, we know that self is a recordset, so instead of doing the intersection python side, where `&` is O(n), but also the number of records returned from a raw `search(domain)` with the domain might be too generic and return a lot of unnecessary records, that we are then discarding python side, you can extend the domain with an `expresson.AND([[('id', 'in', self.ids)], domain])`, so that filtering happens DB side. Should return the same results
- I'm not sure the `with_context(prefetch_fields=False)` does useful anything here?

Reply:
Made the changes!